### ISSUE TITLE
Optionally emulate TTY in Runner

### ISSUE THREAD

**Issue by I-Al-Istannen** [IPDSnelting/velcom#146]
## About
The runner could emulate a TTY ~~(e.g. by calling out to `script` first)~~ to get programs to include ANSI colors, which we can now properly display in the frontend.

`script` does not emulate / handle stderr correctly :/ It seems to reserve stderr for its own output and redirect stderr to stdout for programs it runs. We would need to spawn two PTYs and hook them up to stdout and stderr respectively.

## Implementation
- The runner does not call the bench script directly but wraps its call in a call to ~~`script`~~
- Therefore TTY-Emulation is not performed if the ~~`script`~~ program is not found (if that is used for emulating it)
- Should we add a config toggle in the runner (or backend?) to globally disable TTY emulation?

### Further information
Taken from the comments in #144.

**Comment by I-Al-Istannen** [comment]
I can't find any PTY simulator which simulates *two* PTYs (one for stdout and one for stderr) and passes the output along untangled. `script` only creates one PTY and merges stderr and stdout.

It looks like we'd need to bundle our own PTY emulator script with the runner (e.g. in C or python). This doesn't really sound that desirable or should we assume python or a C compiler is installed on the runner?

**Comment by I-Al-Istannen** [comment]
Something like this seems to suffice (C / Python version):
https://gist.github.com/I-Al-Istannen/3479e22c49aa05bd6f043d6d4660afa8

What's missing is exit code propagation but that should be alright to do.

I am not sure if you'd want to be dependent on python or having a C compiler installed (or the program compiled) for TTY emulation, but it sounds like an acceptable tradeoff. If you don't want python or gcc, you can always just let your bench script wrap its calls in `script` or whatever.

**Comment by Kha** [comment]
Whew. Personally I'm fine with entangling stdout and stderr, so I would be content with the manual "`script` in bench script" solution.

**Comment by I-Al-Istannen** [comment]
I would be fine with it too, but we'd have no way to find out what the actual *result* of the benchmark script is and what is just redirected *output*.

Currently this is separated by stdout = result, stderr = redirected output -- which doesn't work when you only have one stream.

**Comment by Kha** [comment]
Ah, of course. So since the bench script author has to take care of redirecting any secondary output to stderr anyway, it seems fair to have them inject `script`/coloring flags *at that point* as well. Then we get regular stdout and ANSI-encoded secondary stderr.

**Comment by I-Al-Istannen** [comment]
Yea, that was what I meant with
> This patch does not let the runner implement a Pseudo-TTY to fool scripts into outputting colours. If you want colours, you need to write a benchmark script that forces your programs to output colors (e.g. by using --color=always flags or some MAVEN_OPTS and property trickery)

in the PR for the current implementation. As you suggested the runner could emulate the PTY itself, I looked around a bit and the C/Python script is what I came up with. I doubt java can do this natively without resorting to a C module, so we'd need one of those wrappers.

So the question was *whether it is even worth it* to have the runner do that, and it sounds like you agree that it isn't really.

In that case feel free to close this issue. We can always re-open it when there is a more pressing need :)
